src/kafka/protocol/api_versions_response.hh

```diff
@@ -45,6 +54,26 @@ public:
     kafka_array_t<api_versions_response_key> _api_keys;
     kafka_int32_t _throttle_time_ms;
 
+    template<typename RequestType>
+    int16_t max_version() const {
+        auto broker_versions = (*this)[RequestType::API_KEY];
+        if (*broker_versions._api_key == -1) {
+            // Broker does not support specific request.
+            throw unsupported_version_exception();
+        }
+        if (*broker_versions._max_version < RequestType::MIN_SUPPORTED_VERSION) {
+            // Broker is too old.
+            throw unsupported_version_exception();
+        }
+        if (*broker_versions._min_version > RequestType::MAX_SUPPORTED_VERSION) {
+            // Broker is too new.
+            throw unsupported_version_exception();
+        }
+        return std::min(*broker_versions._max_version, RequestType::MAX_SUPPORTED_VERSION);
+    }
+    bool contains(int16_t api_key) const;
+    api_versions_response_key operator[](int16_t api_key) const;
+
     void serialize(std::ostream &os, int16_t api_version) const;
 
     void deserialize(std::istream &is, int16_t api_version);
```

Comment:
This is not regarding this PR, but generally, why does deserialize change the members, but serialize doesn't?

Reply:
Or rather, how

Reply:
Deserialize loads the data from `is` and sets (modifies) object's fields with loaded values. 
Serialize reads the fields of object and writes them to `os` (no modification needed),